pcap/pcapio/ngread.go

```diff
@@ -0,0 +1,317 @@
+}
+
+// DefaultNgReaderOptions provides sane defaults for a pcapng reader.
+var DefaultNgReaderOptions = NgReaderOptions{}
+
+// NgReader wraps an underlying bufio.NgReader to read packet data in pcapng.
+type NgReader struct {
+	*peeker.Reader
+	options       NgReaderOptions
+	ifaces        []NgInterface
+	currentOption ngOption
+	first         []byte
+	ci            gopacket.CaptureInfo
+	bigEndian     bool
+	offset        uint64
+}
+
+// NewNgReader initializes a new writer, reads the first section header, and if necessary according to the options the first interface.
+func NewNgReader(r io.Reader) (*NgReader, error) {
+	ret := &NgReader{
+		Reader: peeker.NewReader(r, 32*1024, 1024*1024),
+	}
+
+	//pcapng _must_ start with a section header
+	typ, block, err := ret.readBlock()
+	if err != nil {
+		return nil, err
+	}
+	if typ != ngBlockTypeSectionHeader {
+		return nil, fmt.Errorf("Unknown magic %x", typ)
+	}
+	if err := ret.parseSectionHeader(block); err != nil {
+		return nil, err
+	}
+	ret.first = block
+	return ret, nil
+}
+
+func (r *NgReader) Packet(block []byte) ([]byte, nano.Ts, layers.LinkType) {
+	//XXX support only enhance packet blocks right now
+	if len(block) < 28 {
+		return nil, 0, 0
+	}
+	ifno := int(r.getUint32(block[8:12]))
+	if ifno >= len(r.ifaces) {
+		return nil, 0, 0
+	}
+	t := time.Unix(r.convertTime(ifno, uint64(r.getUint32(block[12:16]))<<32|uint64(r.getUint32(block[16:20])))).UTC()
+	caplen := int(r.getUint32(block[20:24]))
+	//length := int(r.getUint32(block[24:28]))
+	packet := block[28:]
+	if len(packet) < caplen {
+		return nil, 0, 0
+	}
+	return packet[:caplen], nano.TimeToTs(t), r.ifaces[ifno].LinkType
+}
+
+func (r *NgReader) Offset() uint64 {
+	return r.offset
+}
+
+// The following functions make the binary.* functions inlineable (except for getUint64, which is too big, but not in any hot path anyway)
+// Compared to storing binary.*Endian in a binary.ByteOrder this shaves off about 20% for (ZeroCopy)ReadPacketData, which is caused by the needed itab lookup + indirect go call
+func (r *NgReader) getUint16(buffer []byte) uint16 {
+	if r.bigEndian {
+		return binary.BigEndian.Uint16(buffer)
+	}
+	return binary.LittleEndian.Uint16(buffer)
+}
+
+func (r *NgReader) getUint32(buffer []byte) uint32 {
+	if r.bigEndian {
+		return binary.BigEndian.Uint32(buffer)
+	}
+	return binary.LittleEndian.Uint32(buffer)
+}
+
+func (r *NgReader) getUint64(buffer []byte) uint64 {
+	if r.bigEndian {
+		return binary.BigEndian.Uint64(buffer)
+	}
+	return binary.LittleEndian.Uint64(buffer)
+}
+
+// Now the pcapng implementation
+
+// readBlock reads a the blocktype and length from the file.
+// If the type is a section header, endianess is also read.
+func (r *NgReader) readBlock() (ngBlockType, []byte, error) {
+	hdr, err := r.Peek(8)
+	if err != nil {
+		return 0, nil, err
+	}
+	typ := ngBlockType(r.getUint32(hdr[:4]))
+	length := r.getUint32(hdr[4:8])
+	if length == 0 {
+		// avoid infinite loop for bad input
```

Comment:
You know, I like to point this one out because its a common flaw in protocol design and protocol implementation and thus I think it's worth it to include the comment.   May I reminisce?  I put up a bug fix at Riverbed in the CIFS protocol accelerator module because a length==0 check like this was missing and Kartik closed it with no fix explaining that a sender shouldn't put 0 in the length field.  Ugh.  Van Jacobson taught me years ago that length fields should never include the length of the fixed size header contained within referenced payload so that: (1) you don't have to check for length==0 to avoid an infinite loop (i.e., you can't make this mistake even if you tried) and (2) it lets the payload length be a little longer for the bit budget you have for the length field (which in the old days of TCP/IP header design made some difference but nowadays now big deal).  More to the point, the previous code from gopacket has not quite this bug, but the corollary, where it is computing a length incorrectly and causing an unsigned integer underflow for length value < 8.  The moral of the story is that the  block length field in the pcap-ng spec should have been defined to be the actual length less 12 (to account for the fixed size of the 4-byte block type, block length, and block length trailer) rather than defined as encoding the overall length.  It would be fun to create a pcap-ng file that is fully compliant except for a block that has a length of 0 and feed it to various pcap-related programs and see which ones lock up.  That all said, this discussion made me realize there is a different bug here, which is panic if the length is 1, 2, or 3, but according to Kartik, a file should never have those values here.  Ha!  I will fix this.